Automattic/wp-calypso
# Composite Checkout: Remove useMessages from CheckoutProvider

#### Changes proposed in this Pull Request

The `useMessages` hook in the `@automattic/composite-checkout` package was only being used by one thing: the transaction status handler inside the package. This PR removes the hook entirely to reduce the API size of the package and more importantly to remove three required props on `CheckoutProvider` (`showErrorMessage`, `showInfoMessage`, `showSuccessMessage`) and make one optional (`onPaymentComplete`); removing these props is helpful because the provider is used in several places where those props are never invoked but are still required.

The only downside of this is that the consumer must now handle displaying a notice directly if there is an error or a redirect. To handle these situations, this PR adds two new optional props to `CheckoutProvider`: `onPaymentError` and `onPaymentRedirect`. These compliment the `onPaymentComplete` prop and can be used to display such messages.

The only two places using `CheckoutProvider` currently that would expect to display error and redirect messages are checkout and payment method assignment, so this PR updates each of those implementations to use the new props.

#### Testing instructions

Do not sandbox the API so that using a test credit card number results in an error (we want to see errors).

To test `onPaymentError` in checkout:

- Add a product to your cart and visit checkout.
- Select "Credit or debit card" as the payment method.
- Enter a valid but fake credit card number (eg: `4242 4242 4242 4242`) as the payment method and submit the form.
- Verify that an error message is displayed.

<img width="698" alt="Screen Shot 2021-09-01 at 5 46 06 PM" src="https://user-images.githubusercontent.com/2036909/131749580-1cba0e8e-8ed7-4b9b-b9c0-200b341087f4.png">

To test `onPaymentRedirect` in checkout:

- Add a product to your cart and visit checkout.
- Select PayPal as the payment method and submit the form.
- Verify that just before redirecting, a "Redirecting to payment partner" message is displayed.

<img width="393" alt="Screen Shot 2021-09-01 at 5 46 44 PM" src="https://user-images.githubusercontent.com/2036909/131749642-9e73cac0-b031-4b78-9bd9-958952d11837.png">


To test `onPaymentError` in payment methods:

- Use an account that has at least one active subscription.
- Visit the purchases page for your account at `/me/purchases`.
- Click to select a purchase.
- On the purchase page, click "Add Payment Method" or "Change Payment Method".
- Select "Credit or debit card" as the payment method.
- Enter a valid but fake credit card number (eg: `4242 4242 4242 4242`) as the payment method and submit the form.
- Verify that an error message is displayed.

To test `onPaymentRedirect` in payment methods:

- Use an account that has at least one active subscription.
- Visit the purchases page for your account at `/me/purchases`.
- Click to select a purchase.
- On the purchase page, click "Add Payment Method" or "Change Payment Method".
- Select PayPal as the payment method and submit the form.
- Verify that just before redirecting, a "Redirecting to payment partner" message is displayed.
